### PR TITLE
add CPU only mode which can be activated during install

### PR DESCRIPTION
**Pull request by yossibiton**
This is very useful feature, as it make the repository usable to much more environments. 
For example, i was able to run the webcam demo on my mac using different pretrained models.

### CODE REVIEW

**Comment by yossibiton** [comment]
CI fails because of flake8 style errors - the current issue is with "module level import not at top of file".
This happens because i have blocks like this : 
```
if not mmdet.version.CPU_ONLY:
    from mmdet.ops import MaskedConv2d
```

I didn't find any better alternative because these modules are depend on CUDA and can't be imported in CPU mode. Any suggestions ?

**Comment by peisuke** [comment]
I fixed the flake8 style errors and a bug on GPU. 
I submitted pull request to yossibiton's repository.

File: mmdet/models/anchor_heads/fovea_head.py [hunk 1]
```diff
@@ -2,8 +2,11 @@
 import torch.nn as nn
 from mmcv.cnn import normal_init
 
+import mmdet
 from mmdet.core import multi_apply, multiclass_nms
-from mmdet.ops import DeformConv
+if not mmdet.version.CPU_ONLY:
+    from mmdet.ops import DeformConv
+
```
Comment (xvjiarui) [inline mmdet/models/anchor_heads/fovea_head.py]: We could try to move these kinds of check inside ops like here https://github.com/open-mmlab/mmdetection/blob/dd88ed65131da79bf60bfaf847236ad7c41a6dad/mmdet/ops/dcn/deform_conv.py#L10

**Comment by xvjiarui** [comment]
Hi @yossibiton @peisuke 
Thanks for your contribution. I apologize for the late reply.
May I know the usage of `tools/show_dataset.py`?

File: mmdet/models/anchor_heads/fovea_head.py (hunk 1 above)
Comment (yossibiton) [reply mmdet/models/anchor_heads/fovea_head.py]: i followed your suggestion and did that, please look at my latest commit

**Comment by yossibiton** [comment]
> Hi @yossibiton @peisuke
> Thanks for your contribution. I apologize for the late reply.
> May I know the usage of `tools/show_dataset.py`?

This is a tool i use to look at the training set examples before starting to train. Anyway, i removed it for now, since it's not related to this specific pull request.
I can add it later in a separate pull request if you think it's useful.

File: mmdet/ops/__init__.py [hunk 2]
```diff
@@ -1,15 +1,20 @@
...
+
 from .nms import nms, soft_nms
 from .roi_align import RoIAlign, roi_align
 from .roi_pool import RoIPool, roi_pool
 from .sigmoid_focal_loss import SigmoidFocalLoss, sigmoid_focal_loss
 from .utils import get_compiler_version, get_compiling_cuda_version
 
+if not mmdet.version.CPU_ONLY:
+    from .dcn import (deform_conv, deform_roi_pooling, modulated_deform_conv)
+
```
Comment (xvjiarui) [inline mmdet/ops/__init__.py]: There is no need to check here. I suggest keeping this file unchanged.

File: setup.py [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
+# -*- coding: utf-8 -*-
 import os
+import sys
 import subprocess
 import time
 from setuptools import find_packages, setup
 
 import torch
-from torch.utils.cpp_extension import BuildExtension, CUDAExtension
+from torch.utils.cpp_extension import BuildExtension
```
Comment (xvjiarui) [inline setup.py]: These could be merged.

File: setup.py [hunk 4]
```diff
@@ -92,7 +98,7 @@ def make_cuda_ext(name, module, sources):
     define_macros = []
 
     if torch.cuda.is_available() or os.getenv('FORCE_CUDA', '0') == '1':
-        define_macros += [('WITH_CUDA', None)]
+        define_macros += [("WITH_CUDA", None)]
```
Comment (xvjiarui) [inline setup.py]: The single quote is preferred in this codebase.

**Comment by xvjiarui** [comment]
`CPU_ONLY` could be imported here. 
https://github.com/open-mmlab/mmdetection/blob/f60a63f3589aaa54d4ad142673d9d6697c3c3d73/mmdet/__init__.py#L3

**Comment by xvjiarui** [comment]
Please refer to [contribution guide](https://github.com/open-mmlab/mmdetection/blob/master/.github/CONTRIBUTING.md) to fix the CI error.

File: setup.py [hunk 5]
```diff
@@ -190,47 +202,25 @@ def gen_packages_items():
...
-        extras_require={
-            'all': parse_requirements('requirements.txt'),
-            'tests': parse_requirements('requirements/tests.txt'),
-            'build': parse_requirements('requirements/build.txt'),
-            'optional': parse_requirements('requirements/optional.txt'),
-        },
-        ext_modules=[
-            make_cuda_ext(
+    cpu_only = False
+    if "--cpu" in sys.argv:
```
Comment (xvjiarui) [inline setup.py]: I suggest using `arg_parser `here

File: mmdet/ops/roi_align/roi_align.py [hunk 6]
```diff
@@ -3,7 +3,10 @@
 from torch.autograd.function import once_differentiable
 from torch.nn.modules.utils import _pair
 
-from . import roi_align_cuda
+import mmdet
+
+if not mmdet.version. CPU_ONLY:
```
Comment (xvjiarui) [inline mmdet/ops/roi_align/roi_align.py]: Additional space

Comment (xvjiarui) [reply mmdet/ops/roi_align/roi_align.py]: Please refer to CI output to fix errors.

File: setup.py (hunk 5 above)
Comment (yossibiton) [reply setup.py]: i couldn't use normal arg_parser here because later on distutils expect to get specific args and i didn't want to override them

**Comment by yossibiton** [comment]
@xvjiarui thanks for you review !
i followed your comments and the contributions guide and fixed all issues, except for the arg_parser in setup.py (since i'm nor sure if it's valid to set arg_parser in setup.py)

File: demo/webcam_demo.py [hunk 7]
```diff
@@ -22,8 +23,13 @@ def parse_args():
 def main():
     args = parse_args()
 
+    if mmdet.version.CPU_ONLY:
```
Comment (xvjiarui) [inline demo/webcam_demo.py]: Just `mmdet.CPU_ONLY` will be fine. Similarly for elsewhere.

**Comment by xvjiarui** [comment]
> @xvjiarui thanks for you review !
> i followed your comments and the contributions guide and fixed all issues, except for the arg_parser in setup.py (since i'm nor sure if it's valid to set arg_parser in setup.py)

Maybe we could check if cuda is available to determine which setup to use.

**Comment by yossibiton** [comment]
> > @xvjiarui thanks for you review !
> > i followed your comments and the contributions guide and fixed all issues, except for the arg_parser in setup.py (since i'm nor sure if it's valid to set arg_parser in setup.py)
> 
> Maybe we could check if cuda is available to determine which setup to use.

That's a good idea, so i will use 
`torch.cuda.is_available()`

**Comment by xvjiarui** [comment]
You could refactor this line into a function. 
https://github.com/open-mmlab/mmdetection/blob/6596f5ad448bdbf3f2921c98b43486f096cdd9c6/setup.py#L94

File: setup.py [hunk 8]
```diff
@@ -110,6 +114,12 @@ def make_cuda_ext(name, module, sources):
         })
 
 
+def make_cpp_ext(name, module, sources):
```
Comment (xvjiarui) [inline setup.py]: Could we merge `make_cpp_ext` and `make_cuda_ext` into one function? This would simplify `__main__`.

Comment (yossibiton) [reply setup.py]: did that, please see my new commit

File: setup.py [hunk 9]
```diff
@@ -190,111 +208,130 @@ def gen_packages_items():
...
-            make_cuda_ext(
-                name='compiling_info',
-                module='mmdet.ops.utils',
-                sources=['src/compiling_info.cpp']),
-            make_cuda_ext(
-                name='nms_cpu',
-                module='mmdet.ops.nms',
-                sources=['src/nms_cpu.cpp']),
-            make_cuda_ext(
+    cpu_only = not should_use_cuda()
```
Comment (xvjiarui) [inline setup.py]: It could be deleted. We could call `use_cuda` inside `write_version` and `make_extension`

File: mmdet/datasets/pipelines/transforms.py [hunk 10]
```diff
@@ -638,12 +638,18 @@ def __call__(self, results):
...
+                    new_h = random.uniform(self.min_crop_size * h, h)
+                    new_ar = random.uniform(self.ar_range[0], self.ar_range[1])
+                    new_w = new_h * new_ar
+                else:
+                    new_w = random.uniform(self.min_crop_size * w, w)
+                    new_h = random.uniform(self.min_crop_size * h, h)
+
+                    # h / w in [0.5, 2]
+                    if new_h / new_w < 0.5 or new_h / new_w > 2:
+                        continue
```
Comment (xvjiarui) [inline mmdet/datasets/pipelines/transforms.py]: These could be in another PR.

File: setup.py (hunk 8 above)
Comment (xvjiarui) [reply setup.py]: I suggest determining which extension to use by checking if `.cu` is in the sources? Then there is no need to pass `is_cuda_ext` into function.

File: setup.py [hunk 11]
```diff
@@ -87,27 +90,42 @@ def get_version():
     return locals()['__version__']
 
 
-def make_cuda_ext(name, module, sources):
+def should_use_cuda():
```
Comment (xvjiarui) [inline setup.py]: It could be renamed to just `use_cdua()`.

**Comment by xvjiarui** [comment]
@yossibiton Noted with thanks. 
Could you please resolve the conflicts?

**Comment by yossibiton** [comment]
> @yossibiton Noted with thanks.
> Could you please resolve the conflicts?

just did that. hope everything is ok now.

File: mmdet/datasets/pipelines/transforms.py (hunk 10 above)
Comment (xvjiarui) [reply mmdet/datasets/pipelines/transforms.py]: Could you please keep this part the same as master?

File: setup.py [hunk 12]
```diff
@@ -87,27 +94,45 @@ def get_version():
...
+
 
+def make_extension(name, module, sources, is_cuda_ext=False):
+    name = '{}.{}'.format(module, name)
+    sources = [os.path.join(*module.split('.'), p) for p in sources]
     define_macros = []
 
-    if torch.cuda.is_available() or os.getenv('FORCE_CUDA', '0') == '1':
-        define_macros += [('WITH_CUDA', None)]
+    is_cuda_ext = any([source.endswith('.cu') for source in sources])
```
Comment (xvjiarui) [inline setup.py]: For operators end with `.cpp`, if compiled with cuda it will speed up a little bit.
It would be better if we could keep extensions unchanged when `CPU_ONLY==False`.

File: docs/INSTALL.md [hunk 13]
```diff
@@ -60,6 +60,25 @@ you can install it before installing MMCV.
 
 4. Some dependencies are optional. Simply running `pip install -v -e .` will only install the minimum runtime requirements. To use optional dependencies like `albumentations` and `imagecorruptions` either install them manually with `pip install -r requirements/optional.txt` or specify desired extras when calling `pip` (e.g. `pip install -v -e .[optional]`). Valid keys for the extras field are: `all`, `tests`, `build`, and `optional`.
 
+## Install in CPU_ONLY mode
+The code can be built for CPU only environment (where CUDA isn't available) by using the install command :
+```shell
+python setup.py develop --cpu
+```
```
Comment (xvjiarui) [inline docs/INSTALL.md]: These are outdated.

File: docs/INSTALL.md [hunk 14]
```diff
@@ -60,6 +60,25 @@ you can install it before installing MMCV.
...
+After installation, you can always get the current mode by checking the constant : ```mmdet.version.CPU_ONLY```
```
Comment (xvjiarui) [inline docs/INSTALL.md]: Change to `mmdet.CPU_ONLY`.

**Comment by xvjiarui** [comment]
Hi @yossibiton 
Thanks for the fix. We are having a refactor recently. I will get back to you very soon.

**Comment by xvjiarui** [comment]
Hi @yossibiton ,
I opened a new PR #2385 on this feature. You are very welcomed to contribute to that PR.

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @yossibiton ！We are grateful for your efforts in helping improve mmdetection open-source project during your personal time. 

Welcome to join OpenMMLab Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA
If you have a WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）

Thank you again for your contribution❤
